### PR TITLE
[flang] Enhance location information

### PR DESCRIPTION
**Pull request by clementval**
Add inclusion location information by using FusedLocation with attribute. 

More context here: https://discourse.llvm.org/t/rfc-enhancing-location-information/79650

### CODE REVIEW

**Comment by abidh** [comment]
The if condition [here](https://github.com/llvm/llvm-project/blob/ca7dc895cea44c80263c969302fa2f202751aa9c/flang/lib/Optimizer/Transforms/AddDebugInfo.cpp#L231) will need to be adjusted otherwise we will drop all the debug information inside the include files.

**Comment by clementval** [comment]
> The if condition [here](https://github.com/llvm/llvm-project/blob/ca7dc895cea44c80263c969302fa2f202751aa9c/flang/lib/Optimizer/Transforms/AddDebugInfo.cpp#L231) will need to be adjusted otherwise we will drop all the debug information inside the include files.

Sure I'll adapt that.

**Review by vzakhari — COMMENTED** [review]
Thank you, Valentin!  It looks good to me, but I think we'd better use a more specialized attribute to represent the locations fused due to source inlining.  The integer attribute seems to be too generic to me.

**Comment by clementval** [comment]
> Thank you, Valentin! It looks good to me, but I think we'd better use a more specialized attribute to represent the locations fused due to source inlining. The integer attribute seems to be too generic to me.

Yes, I think a proper attribute will be better to have since it might need to be expanded later.

File: flang/test/Lower/location.f90 [hunk 1]
```diff
@@ -0,0 +1,11 @@
+! RUN: bbc -emit-hlfir --mlir-print-debuginfo %s -o - | FileCheck %s
```
Comment (jeanPerier) [inline flang/test/Lower/location.f90:1]: This test seems to fail in the GitHub CI.

**Review by jeanPerier — APPROVED** [review]
Please check the CI failure with the added test, LGTM otherwise.

**Comment by jeanPerier** [comment]
> > The if condition [here](https://github.com/llvm/llvm-project/blob/ca7dc895cea44c80263c969302fa2f202751aa9c/flang/lib/Optimizer/Transforms/AddDebugInfo.cpp#L231) will need to be adjusted otherwise we will drop all the debug information inside the include files.
> 
> Sure I'll adapt that.

Did you adapt it in this patch, I do not see a change in AddDebugInfo.cpp here.

**Comment by clementval** [comment]
> > > The if condition [here](https://github.com/llvm/llvm-project/blob/ca7dc895cea44c80263c969302fa2f202751aa9c/flang/lib/Optimizer/Transforms/AddDebugInfo.cpp#L231) will need to be adjusted otherwise we will drop all the debug information inside the include files.
> > 
> > 
> > Sure I'll adapt that.
> 
> Did you adapt it in this patch? I do not see a change in AddDebugInfo.cpp in this patch.

No I still need to update it.

File: flang/lib/Optimizer/Transforms/AddDebugInfo.cpp [hunk 2]
```diff
@@ -76,7 +76,7 @@ static uint32_t getLineFromLoc(mlir::Location loc) {
 }
 
 bool debugInfoIsAlreadySet(mlir::Location loc) {
-  if (mlir::isa<mlir::FusedLoc>(loc))
+  if (l->findInstanceOf<mlir::FusedLocWith<mlir::LLVM::DISubprogramAttr>>())
```
Comment (clementval) [inline flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:79]:
```suggestion
  if (loc->findInstanceOf<mlir::FusedLocWith<mlir::LLVM::DISubprogramAttr>>())
```

File: flang/test/Lower/location.f90 [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+program test
+include 'location0.inc'
+
+end 
+
+! CHECK-LABEL: func.func @_QQmain() attributes {fir.bindc_name = "test"} {
+! CHECK: fir.call @_FortranAioOutputAscii(%{{.*}}, %{{.*}}, %{{.*}}) fastmath<contract> : (!fir.ref<i8>, !fir.ref<i8>, i64) -> i1 loc(fused<#fir<loc_kind_array[ base,  inclusion,  inclusion]>>["{{.*}}flang/test/Lower/location1.inc":1:10, "{{.*}}flang/test/Lower/location0.inc":1:1, "{{.*}}flang/test/Lower/location.f90":4:1])
+! CHECK: return loc("{{.*}}flang/test/Lower/location.f90":6:1)
+! CHECK: } loc("{{.*}}flang/test/Lower/location.f90":3:1)
```
Comment (clementval) [inline flang/test/Lower/location.f90:11]:
```suggestion
! CHECK: fir.call @_FortranAioOutputAscii(%{{.*}}, %{{.*}}, %{{.*}}) fastmath<contract> : (!fir.ref<i8>, !fir.ref<i8>, i64) -> i1 loc(fused<#fir<loc_kind_array[ base,  inclusion,  inclusion]>>["{{.*}}location1.inc":1:10, "{{.*}}location0.inc":1:1, "{{.*}}flang/test/Lower/location.f90":4:1])
! CHECK: return loc("{{.*}}location.f90":6:1)
! CHECK: } loc("{{.*}}location.f90":3:1)
```

File: flang/test/Lower/location.f90 [hunk 4]
```diff
@@ -0,0 +1,13 @@
+! RUN: bbc -emit-hlfir --mlir-print-debuginfo %s -o - | FileCheck %s
+
+program test
+include 'location0.inc'
+
+end 
+
+! CHECK-LABEL: func.func @_QQmain() attributes {fir.bindc_name = "test"} {
+! CHECK: fir.call @_FortranAioOutputAscii(%{{.*}}, %{{.*}}, %{{.*}}) fastmath<contract> : (!fir.ref<i8>, !fir.ref<i8>, i64) -> i1 loc(fused<#fir<loc_kind_array[ base,  inclusion,  inclusion]>>["{{.*}}location1.inc":1:10, "{{.*}}location0.inc":1:1, "{{.*}}flang/test/Lower/location.f90":4:1])
```
Comment (clementval) [inline flang/test/Lower/location.f90:9]:
```suggestion
! CHECK: fir.call @_FortranAioOutputAscii(%{{.*}}, %{{.*}}, %{{.*}}) fastmath<contract> : (!fir.ref<i8>, !fir.ref<i8>, i64) -> i1 loc(fused<#fir<loc_kind_array[ base,  inclusion,  inclusion]>>["{{.*}}location1.inc":1:10, "{{.*}}location0.inc":1:1, "{{.*}}location.f90":4:1])
```

**Comment by clementval** [comment]
@vzakhari Are you ok with the newly created attribute? Or were you thinking about something different?

**Comment by vzakhari** [comment]
> @vzakhari Are you ok with the newly created attribute? Or were you thinking about something different?

Yes, it looks good to me.  Thank you, Valentin!

**Review by vzakhari — APPROVED** [review]

**Pull request merged**